OpenRA.Mods.D2k/Traits/Carryable.cs

```diff
@@ -17,24 +17,49 @@
 namespace OpenRA.Mods.D2k.Traits
 {
 	[Desc("Can be carried by units with the trait `Carryall`.")]
-	public class CarryableInfo : ITraitInfo
+	public class CarryableInfo : ITraitInfo, Requires<UpgradeManagerInfo>
 	{
 		[Desc("Required distance away from destination before requesting a pickup. Default is 6 cells.")]
-		public WDist MinDistance = WDist.FromCells(6);
+		public readonly WDist MinDistance = WDist.FromCells(6);
+
+		[UpgradeGrantedReference]
+		[Desc("The upgrades to grant to self while waiting or being carried.")]
+		public readonly string[] CarryableUpgrades = { };
```

Comment:
This makes the `WithDecorationCarryable` trait redundant.  You can now use `WithDecoration` enabled by this upgrade.